### PR TITLE
issue 235: kafka consumer and publisher for multiple brokers

### LINKED ISSUE #235: Issue with kafka brokers
I have a setup with 4 kafka brokers, and 4 partitions per `gobmp` topic. All topics are pre-created, replication factor is set to 4. However, in gobmp `kafka-server` variable it seems I cannot provide all the brokers, but it accepts only one https://github.com/sbezverk/gobmp/blob/master/pkg/kafka/kafka-publisher.go#L206
Which is not a correct practice in kafka, since it needs all broker endpoints to be provided for the process to work correctly.
Can the `kafka-server` support the typical use case, which is a comma-separated list of brokers?
Note that if I simply supply only one broker, sarama complains underneath
```
gobmp.go:74] failed to initialize Kafka publisher with error: kafka server: This is not the correct controller for this cluster.
```
because of course any of the 4 brokers may be the leader at any time. 
Could you help me with a refactoring of the validation process for kafka endpoints please?

### PR DESCRIPTION
**Pull request by kotronis-te**
This PR adds support for multiple brokers using sarama ClusterAdmin interface.
It includes (currently in draft) docker-compose files to initiate a local installation with a kafka cluster and zookeeper.

Closes #235

### CODE REVIEW

File: pkg/kafka/kafka-publisher.go [hunk 1]
```diff
@@ -160,21 +161,26 @@ func NewKafkaPublisher(kafkaSrv string) (pub.Publisher, error) {
 	config.Admin.Retry.Max = 100
 	config.Version = sarama.V1_1_0_0
 
-	br := sarama.NewBroker(kafkaSrv)
+	kafkaSrvs := strings.Split(kafkaSrv, ",")
+	ca, err := sarama.NewClusterAdmin(kafkaSrvs, config)
+	if err != nil {
+		glog.Errorf("failed to create cluster admin: %+v", err)
```
Comment (sbezverk) [inline pkg/kafka/kafka-publisher.go:171]: Are you sure it is safe to just print error if NewClusterAdmin fails? I see ca is used as a parameter in waitForControllerBrokerConnection.

File: pkg/kafka/kafka-publisher.go [hunk 2]
```diff
@@ -260,34 +264,38 @@ func ensureTopic(br *sarama.Broker, timeout time.Duration, topicName string) err
...
 
-func waitForBrokerConnection(br *sarama.Broker, config *sarama.Config, timeout time.Duration) error {
+func waitForControllerBrokerConnection(ca *sarama.ClusterAdmin, config *sarama.Config, timeout time.Duration) (cb *sarama.Broker, error) {
 	ticker := time.NewTicker(10 * time.Second)
 	tout := time.NewTimer(timeout)
 	defer func() {
 		ticker.Stop()
 		tout.Stop()
 	}()
+	cb, err := ca.Controller()
```
Comment (sbezverk) [inline pkg/kafka/kafka-publisher.go:282]: See previous comment, is ca is nil, you will get panic here.

**Review by sbezverk — CHANGES_REQUESTED** [review]
Please fix possible panic.

File: pkg/kafka/kafka-publisher.go (hunk 1 above)
Comment (kotronis-te) [reply pkg/kafka/kafka-publisher.go:171]: fixed in https://github.com/sbezverk/gobmp/pull/236/commits/3c7fbf3dbe99ec0ef3981dcc8634c1931cb1c054

File: pkg/kafka/kafka-publisher.go (hunk 2 above)
Comment (kotronis-te) [reply pkg/kafka/kafka-publisher.go:282]: fixed in https://github.com/sbezverk/gobmp/pull/236/commits/3c7fbf3dbe99ec0ef3981dcc8634c1931cb1c054

File: Makefile [hunk 3]
```diff
@@ -1,5 +1,5 @@
 REGISTRY_NAME?=docker.io/sbezverk
-IMAGE_VERSION?=0.0.0
+IMAGE_VERSION?=test-235
```
Comment (kotronis-te) [inline Makefile]: this I will revert (used only for temp testing to make the container)

File: go.mod [hunk 4]
```diff
@@ -3,7 +3,7 @@ module github.com/sbezverk/gobmp
 go 1.19
 
 require (
-	github.com/Shopify/sarama v1.27.0
+	github.com/IBM/sarama v1.42.1
```
Comment (kotronis-te) [inline go.mod:6]: I updated this to the proper most recent version

**Comment by kotronis-te** [comment]
Hmmm I will turn to undraft when I get messages to kafka. I added sample prefixes to the XR's but I cannot seem to get sth in the single- or multi-broker case even with the initial gobmp container, looking at `gobmp.parsed.unicast_prefix_v4` with console consumer. Any ideas on this?

**Comment by sbezverk** [comment]
I checked the existing pipeline and it seems to be ok, I have another PR in flight, all checks are green. I think there is something with connectivity with new test bed.
```
E1128 20:11:45.006677       1 kafka-publisher.go:168] failed to create cluster admin: kafka: client has run out of available brokers to talk to: dial tcp 10.1.1.4:9092: connect: connection refused
E1128 20:11:45.006785       1 gobmp.go:84] failed to initialize Kafka publisher with error: kafka: client has run out of available brokers to talk to: dial tcp 10.1.1.4:9092: connect: connection refused
```

Also XRd validator does not see BGP sessions coming up and BMP session coming up. I will try to use your docker compose to bring it up and check..

**Comment by kotronis-te** [comment]
> @kotronis-te The issue was with kafka client/metadata configuration, see my test PR for diff.

I am also getting the following on ios xr side:
```
xr-1  | RP/0/RP0/CPU0:Nov 29 08:33:14.447 UTC: smartlicserver[123]: %LICENSE-SMART_LIC-6-EXPORT_CONTROLLED : Usage of export controlled features is not allowed
xr-1  | RP/0/RP0/CPU0:Nov 29 08:33:14.479 UTC: smartlicserver[123]: %LICENSE-SMART_LIC-3-COMM_FAILED : Communications failure with the Cisco Smart Software Manager (CSSM) : Communications init failure
```

**Comment by kotronis-te** [comment]
Hmmm I think I am on to sth, the admin seems to want to create aux topics that have replication factor of 3 (minimum), which will not work in the single-broker case. Checking and will come back to this.
```
testbed-kafka-1-1  | [2023-11-29 08:44:04,397] INFO [Admin Manager on Broker 1]: Error processing create topic request CreatableTopic(name='__consumer_offsets', numPartitions=50, replicationFactor=3, assignments=[], configs=[CreateableTopicConfig(name='compression.type', value='producer'), CreateableTopicConfig(name='cleanup.policy', value='compact'), CreateableTopicConfig(name='segment.bytes', value='104857600')]) (kafka.server.ZkAdminManager)
testbed-kafka-1-1  | org.apache.kafka.common.errors.InvalidReplicationFactorException: Replication factor: 3 larger than available brokers: 1.
```
In the meanwhile, I will integrate your changes.

**Comment by kotronis-te** [comment]
config updates in https://github.com/sbezverk/gobmp/pull/236/commits/3c7fbf3dbe99ec0ef3981dcc8634c1931cb1c054
kafka offset fix in https://github.com/sbezverk/gobmp/pull/236/commits/747c983d71b7717e2d7677ef9d7b8607e24dab65

Now the main issue that remains is with the XRs, will consider using a different router like gobgp to test the setup.

**Comment by sbezverk** [comment]
@kotronis-te you can safely ignore these:
```
xr-1  | RP/0/RP0/CPU0:Nov 29 08:33:14.447 UTC: smartlicserver[123]: %LICENSE-SMART_LIC-6-EXPORT_CONTROLLED : Usage of export controlled features is not allowed
xr-1  | RP/0/RP0/CPU0:Nov 29 08:33:14.479 UTC: smartlicserver[123]: %LICENSE-SMART_LIC-3-COMM_FAILED : Communications failure with the Cisco Smart Software Manager (CSSM) : Communications init failure
```

we are using xrd in eval mode.

**Comment by kotronis-te** [comment]
> @kotronis-te you can safely ignore these:
> 
> ```
> xr-1  | RP/0/RP0/CPU0:Nov 29 08:33:14.447 UTC: smartlicserver[123]: %LICENSE-SMART_LIC-6-EXPORT_CONTROLLED : Usage of export controlled features is not allowed
> xr-1  | RP/0/RP0/CPU0:Nov 29 08:33:14.479 UTC: smartlicserver[123]: %LICENSE-SMART_LIC-3-COMM_FAILED : Communications failure with the Cisco Smart Software Manager (CSSM) : Communications init failure
> ```
> 
> we are using xrd in eval mode.

OK will ping when I switch the testbed I use to gobgp which I had tested also in the past

**Comment by sbezverk** [comment]
Not sure if it is worth it to switch to gobgp, as your change with a single node kafka is green, all expected messages for unicast v4 are received. Have nothing against using gobgp in future in ci/cd pipeline, just now let's try to keep less changes.

**Comment by kotronis-te** [comment]
> Not sure if it is worth it to switch to gobgp, as your change with a single node kafka is green, all expected messages for unicast v4 are received. Have nothing against using gobgp in future in ci/cd pipeline, just now let's try to keep less changes.

No worries it will be a separate compose and configuration, used for testing. Will ping soon when tested.

**Comment by kotronis-te** [comment]
Fixed in https://github.com/sbezverk/gobmp/pull/236/commits/49044c99a6af9843530d43d7a0bff51ce9dce2d3
seems to work like a charm
```
{"action":"add","router_hash":"87d86e0d3cf65f69738c696c5ab800b4","router_ip":"10.1.1.4","base_attrs":{"base_attr_hash":"37ec0eb8db18c3725ef01c705b9b9c85","origin":"incomplete","as_path":[1111],"as_path_count":1,"nexthop":"10.1.1.3","is_atomic_agg":false},"peer_hash":"934b7a4989b7ac8024ddaf4c50cd3185","peer_ip":"10.1.1.3","peer_type":0,"peer_asn":1111,"timestamp":"2023-11-29T09:40:40Z","prefix":"192.168.3.0","prefix_len":24,"is_ipv4":true,"origin_as":1111,"nexthop":"10.1.1.3","is_nexthop_ipv4":true,"is_adj_rib_in_post_policy":false,"is_adj_rib_out_post_policy":false,"is_loc_rib_filtered":false}
```

Opening the PR for review, let me know what to edit when available.

File: Makefile (hunk 3 above)
Comment (kotronis-te) [reply Makefile]: reverted in https://github.com/sbezverk/gobmp/pull/236/commits/43f15de7a282824ce0fbc241654ca299fb6dbe40

File: pkg/kafka/testbed/docker-compose-multi-broker-gobgp.yaml [hunk 5]
```diff
@@ -0,0 +1,190 @@
+version: "3.2"
```
Comment (sbezverk) [inline pkg/kafka/testbed/docker-compose-multi-broker-gobgp.yaml]: docker compose files should not be in pkg/kafka

Please see how the current ci/cd pipeline is built and try to add 4 nodes cluster as another integration test.

File: pkg/kafka/testbed/docker-compose-multi-broker-gobgp.yaml [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+  gobmp-gobgp:
+    driver: bridge
+    ipam:
+      driver: default
+      config:
+      - subnet: 10.1.1.0/24
+services:
+  gobmp:
+    #image: sbezverk/gobmp:v1.0.3
+    image: sbezverk/gobmp:test-235
```
Comment (sbezverk) [inline pkg/kafka/testbed/docker-compose-multi-broker-gobgp.yaml]: image tag should be `:cicd` as that is how gobmp container is built for testing.

File: pkg/kafka/testbed/docker-compose-multi-broker-gobgp.yaml [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+    volumes:
+      - ./zookeeper_hc.sh:/tmp/hc.sh
+    healthcheck:
+      test: [ "CMD-SHELL", "/tmp/hc.sh" ]
+      interval: 2s
+      timeout: 5s
+      retries: 5
+      start_period: 20s
+  kafka-1:
+    image: confluentinc/cp-kafka:7.3.1
```
Comment (sbezverk) [inline pkg/kafka/testbed/docker-compose-multi-broker-gobgp.yaml]: why switch to confluentinc/cp-kafka:7.3.1 ?

File: pkg/kafka/testbed/docker-compose-multi-broker-gobgp.yaml [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+    restart: always
+    depends_on:
+      - zookeeper
+    expose:
+      - 9091
+    environment:
+      KAFKA_BROKER_ID: 1
+      KAFKA_ZOOKEEPER_CONNECT: "zookeeper:2181"
+      ZOOKEEPER: "zookeeper:2181"
+      KAFKA_TOPICS: "gobmp.parsed.peer:4:4:delete gobmp.parsed.unicast_prefix:6:4:delete gobmp.parsed.unicast_prefix_v4:6:4:delete gobmp.parsed.unicast_prefix_v6:6:4:delete gobmp.parsed.ls_node:4:4:delete gobmp.parsed.ls_link:4:4:delete gobmp.parsed.l3vpn:4:4:delete gobmp.parsed.l3vpn_v4:4:4:delete gobmp.parsed.l3vpn_v6:4:4:delete gobmp.parsed.sr_policy:4:4:delete gobmp.parsed.sr_policy_v4:4:4:delete gobmp.parsed.sr_policy_v6:4:4:delete gobmp.parsed.flowspec:4:4:delete gobmp.parsed.flowspec_v4:4:4:delete gobmp.parsed.flowspec_v6:4:4:delete gobmp.parsed.ls_prefix:4:4:delete gobmp.parsed.ls_srv6_sid:4:4:delete gobmp.parsed.evpn:4:4:delete gobmp.parsed.statistics:4:4:delete"
```
Comment (sbezverk) [inline pkg/kafka/testbed/docker-compose-multi-broker-gobgp.yaml]: why do you create topics here? the code will create a topic on a needed basis.

File: pkg/kafka/testbed/xrd-2_xrconf.cfg [hunk 9]
```diff
@@ -0,0 +1,32 @@
+hostname r2
```
Comment (sbezverk) [inline pkg/kafka/testbed/xrd-2_xrconf.cfg]: no need to have xr configs here

File: pkg/kafka/testbed/gobgp/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,21 @@
+FROM alpine:3.16.3
```
Comment (sbezverk) [inline pkg/kafka/testbed/gobgp/Dockerfile]: move all gobgp related changes in a separate PR once you complete multi node pipeline with xrd

**Review by sbezverk — CHANGES_REQUESTED** [review]

**Comment by kotronis-te** [comment]
@sbezverk I will turn to draft and work on the restructuring you propose so that this is better integrated with the CI.

File: pkg/kafka/testbed/docker-compose-multi-broker-gobgp.yaml (hunk 5 above)
Comment (kotronis-te) [reply pkg/kafka/testbed/docker-compose-multi-broker-gobgp.yaml]: working on this. already moved the files to correct location

File: pkg/kafka/testbed/docker-compose-multi-broker-gobgp.yaml (hunk 6 above)
Comment (kotronis-te) [reply pkg/kafka/testbed/docker-compose-multi-broker-gobgp.yaml]: fixed

File: pkg/kafka/testbed/docker-compose-multi-broker-gobgp.yaml (hunk 7 above)
Comment (kotronis-te) [reply pkg/kafka/testbed/docker-compose-multi-broker-gobgp.yaml]: switch reverted, using the testbed image you already use

File: pkg/kafka/testbed/docker-compose-multi-broker-gobgp.yaml (hunk 8 above)
Comment (kotronis-te) [reply pkg/kafka/testbed/docker-compose-multi-broker-gobgp.yaml]: reverted, using auto-creation

File: pkg/kafka/testbed/xrd-2_xrconf.cfg (hunk 9 above)
Comment (kotronis-te) [reply pkg/kafka/testbed/xrd-2_xrconf.cfg]: moved to correct place

File: pkg/kafka/testbed/gobgp/Dockerfile (hunk 10 above)
Comment (kotronis-te) [reply pkg/kafka/testbed/gobgp/Dockerfile]: will do, not relevant to this PR then

**Comment by kotronis-te** [comment]
I am facing some issues with the current CI, will un-draft when resolved.

**Comment by kotronis-te** [comment]
@sbezverk I am running the CI/CD commands exactly as they are in the GH Action workflow:
```
    make container REGISTRY_NAME=localhost IMAGE_VERSION=cicd
    make validator
    docker pull sbezverk/xrd-control-plane:7.9.2
    docker compose -f ./build/unicastv4-bgp/test_bed.yml up --detach
   ./build/monitor_container.sh gobmp
   ./bin/validator --validate=true --kafka=127.0.0.1:9092 --msg-file=./testdata/validator/ipv4_unicast.msg
```
however even for the classic workflow I get:
```
I1220 14:20:59.326701   68933 kafka-consumer.go:35] NewKafkaConsumer
I1220 14:20:59.330829   68933 check.go:39] For message type 74, 3 messages found
I1220 14:20:59.331051   68933 check.go:76] Dictionary for topic type 74 contains 3 test messages
E1220 14:21:09.331936   68933 validator.go:96] timed out waiting for the test to complete
```

Any ideas? You had said at one point that you had verified the CI with the new changes, can you let me know what you did please and how the timeout was remedied?

This is blocking from testing also the other new workflows I have added, since I still get the timeout

**Review by sbezverk — APPROVED** [review]

**Comment by sbezverk** [comment]
@kotronis-te  from the first look, it seems ok, but we need to test it in gihub cicd, try to add this code at the end  of `cicd.yml`:

```yaml
      - name: stop unicast v4 integration test
         run: docker compose -f ./build/unicastv4-bgp/test_bed.yml down     
      - name: starts kafka multi broker
        run: docker compose -f ./build/kafka-single-multi-broker/test_bed_multi_broker.yaml up --detach
      - name: check conection to kafka and bmp session with a router
        id: session2
        run: ./build/monitor_container.sh gobmp
        continue-on-error: true
      - name: Check on failures
        if: steps.session2.outcome != 'success'
        run: |
          docker ps -a
          docker logs gobmp
          docker logs xr-1
          exit 1 
      - name: stop kafka multi broker
        run: docker compose -f ./build/kafka-single-multi-broker/test_bed_multi_broker.yaml down
      - name: starts kafka single broker
        run: docker compose -f ./build/kafka-single-multi-broker/test_bed_one_broker.yaml up --detach
      - name: check conection to kafka and bmp session with a router
        id: session3
        run: ./build/monitor_container.sh gobmp
        continue-on-error: true
      - name: Check on failures
        if: steps.session3.outcome != 'success'
        run: |
          docker ps -a
          docker logs gobmp
          docker logs xr-1
          exit 1 
      - name: stop kafka single broker
        run: docker compose -f ./build/kafka-single-multi-broker/test_bed_one_broker.yaml down
```

Let's see if your scripts work in cicd.

**Comment by kotronis-te** [comment]
Closing, will re-initiate PR from other official repo for better software management.
